### ISSUE TITLE
Exception thrown trying to deserialize native types (e.g. int)

### ISSUE THREAD

**Issue by mattvryan** [FasterXML/jackson-dataformat-xml#139]
I'm seeing a JsonMappingException when I try to deserialize a string into an int.  The exception message says, "Can not deserialize instance of java.lang.Integer out of START_OBJECT token".  I'm using version 2.5.1 for jackson-core, jackson-databind, jackson-annotations, and jackson-dataformat-xml.  I'm also using woodstox-core-asl 4.4.1 although it doesn't seem to make any difference if I use this or not.

Is this behavior supported?

I have a simple JUnit unit test that reproduces the behavior:
```java
    @Test
    public void testXmlDeserializeInt() throws IOException
    {
        int i = 111;
        ObjectMapper mapper = new XmlMapper();
        String s = mapper.writeValueAsString(i);
        int i2 = mapper.readValue(s, int.class);
        assertEquals(i, i2);
    }
```
The exception is thrown on the `readValue()` call.  The String value being passed in is `<Integer>111</Integer>`.  You can see from the code that the string I'm trying to deserialize is the same one created by the mapper to serialize the value in the first place.  Thus the behavior appears to not be symmetric.

Using the Integer type fails similarly:
```java
    @Test
    public void testXmlDeserializeInt() throws IOException
    {
        Integer i = 111;
        ObjectMapper mapper = new XmlMapper();
        String s = mapper.writeValueAsString(i);
        Integer i2 = mapper.readValue(s, Integer.class);
        assertEquals(i, i2);
    }
```
The only apparent difference is the exception message - it says it cannot deserialize an instance of int, instead of java.lang.Integer.

If I use a JSON mapper instead of an XML mapper this test works fine (whether I use int or Integer doesn't matter):
```java
    @Test
    public void testXmlDeserializeInt() throws IOException
    {
        int i = 111;
        ObjectMapper mapper = new ObjectMapper();
        String s = mapper.writeValueAsString(i);
        int i2 = mapper.readValue(s, int.class);
        assertEquals(i, i2);
    }
```

The full stack trace is:
```
com.fasterxml.jackson.databind.JsonMappingException: Can not deserialize instance of java.lang.Integer out of START_OBJECT token
 at [Source: java.io.StringReader@48c76607; line: 1, column: 1]
	at com.fasterxml.jackson.databind.JsonMappingException.from(JsonMappingException.java:148)
	at com.fasterxml.jackson.databind.DeserializationContext.mappingException(DeserializationContext.java:835)
	at com.fasterxml.jackson.databind.deser.std.StdDeserializer._parseInteger(StdDeserializer.java:429)
	at com.fasterxml.jackson.databind.deser.std.NumberDeserializers$IntegerDeserializer.deserialize(NumberDeserializers.java:289)
	at com.fasterxml.jackson.databind.deser.std.NumberDeserializers$IntegerDeserializer.deserialize(NumberDeserializers.java:271)
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:3562)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2578)
	at com.zoomulus.weaver.rest.RestServerTest.testXmlDeserializeInt(RestServerTest.java:1688)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:50)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:459)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:675)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:382)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:192)
```

**Comment by cowtowncoder** [comment]
Hmmh. I wouldn't say result is as intended, but I would say that ability to serialize scalar types as root values is not explicitly supported. So I would recommend only serializing POJOs as root values (and not scalars, arrays or Lists); properties can obviously have any values.

Having said that if it is possible to improve things so that this would work, I'd be happy to improve it.
So I'll consider this a possible bug, with caveat that in some cases it may not be possible to improve handling.
If it turns out not to be possible I'll update README to further state the limitation.

Thank you for reporting this.

**Comment by mattvryan** [comment]
@cowtowncoder Thanks for considering the issue.  I had read that serialization/deserialization of scalar types as root values wasn't explicitly supported, but when it would serialize to XML I figured maybe that information was outdated.

Would the java.lang.Integer class be considered a POJO or a scalar type?

**Comment by cowtowncoder** [comment]
That would be scalar type in this context (results in scalar JSON value; is not accessed via property accessors).

**Comment by mattvryan** [comment]
Thanks for the information.

Here's some brief context:  I'm writing a REST framework that, under certain conditions, can make accommodations for the content type of a message based on context.  In this particular example, suppose the endpoint is declared as a POST endpoint with a single POJO argument.  If the content-type of the request is application/xml, my framework will automatically try to use an XmlMapper() to convert the POST payload into an object of the matching type to pass as the argument to the endpoint.
The problem, which led us to this issue, is if the argument is a scalar type the conversion does not work.  However, it does work for scalar types if the request content-type is application/json.

In short this puts me in a situation where support for application/xml is not as robust as application/json, which I can work around but I'd like them to be similar.

I'm happy to help.  I'll fork the code and look into a fix.

**Comment by cowtowncoder** [comment]
Sounds good. Always happy to get help in any form!

FWIW, JSON specification itself does not consider root-level non-Object/Array values to be valid JSON, so while Jackson handles this without problems, it is not necessarily a robust things to support.
So it is not necessarily unreasonable to state that content to upload must be an Object (and/or Array).

**Comment by mattvryan** [comment]
I've been looking into a possible fix but I must admit I'm unsure how to proceed.  I believe the root of this problem is the difference in how scalar values are represented between the two formats.  In JSON the integer value 123 is represented simply as a string, e.g. "123".  In XML this is represented as `<Integer>123</Integer>`.  This causes the tokenizer to lead off with a `START_OBJECT` token in XML which doesn't happen in JSON.

A fix that appears to work is to modify the StdDeserializer in jackson-databind to tolerate these unexpected tokens during parsing.  For example, changing the first few lines of `StdDeserializer._parseInteger()` from:
```java
        JsonToken t = jp.getCurrentToken();
        if (t == JsonToken.VALUE_NUMBER_INT || t == JsonToken.VALUE_NUMBER_FLOAT) { // coercing should work too
            return Integer.valueOf(jp.getIntValue());
        }
```
to
```java
        JsonToken t = jp.getCurrentToken();
        
        while (null != t && (t == JsonToken.START_OBJECT || t == JsonToken.FIELD_NAME)) t = jp.nextToken();
        
        if (t == JsonToken.VALUE_NUMBER_INT || t == JsonToken.VALUE_NUMBER_FLOAT) { // coercing should work too
            return Integer.valueOf(jp.getIntValue());
        }
```
passes my unit test; additionally all other deserialization unit tests in both jackson-databind and jackson-dataformat-xml that were previously passing continue to pass.

Of course that only works for Integer and int; similar changes would have to be made for other scalar types.

I assume it would be preferable to isolate the changes to this project rather than make changes in jackson-databind.  That's why I'm unsure how to proceed.

Any ideas?

**Comment by cowtowncoder** [comment]
Right, I would rather not have format-level work-arounds at databinding level. This why I am not big fan of this approach.

However, given that this only affects "root values", it may be possible to handle this at slightly different place. You may want to look at interaction of mapper and `DeserializationContext` in xml module; there are some other things that get initialized differently I think.

**Comment by mattvryan** [comment]
I created pull request #140 which implements a potential fix.  The change is isolated to FromXmlParser.  Interested to know what you think.  All prior deserialization unit tests still pass along with a number of new tests added for testing scalar deserialization.

**Comment by mattvryan** [comment]
Per concerns raised in #140 which I agree with, I'm closing this
